exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,624 @@
+/*
+ * Copyright 2019 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use exonum::{
+    api::ApiContext,
+    blockchain::Schema as CoreSchema,
+    crypto::{Hash, PublicKey, SecretKey},
+    helpers::{Height, ValidatorId},
+    messages::BinaryValue,
+    exonum_merkledb::{self, Fork, Snapshot},
+    node::ApiSender,
+    proto::Any,
+    runtime::{
+        api::ServiceApiBuilder,
+        dispatcher::{Dispatcher, DispatcherSender},
+        ArtifactId, ArtifactProtobufSpec, CallInfo, ErrorKind, ExecutionContext, ExecutionError,
+        InstanceDescriptor, InstanceId, InstanceSpec, Runtime, RuntimeIdentifier,
+        StateHashAggregator,
+    },
+};
+use futures::{Future, IntoFuture};
+use jni::{
+    objects::{GlobalRef, JObject, JValue},
+    signature::{JavaType, Primitive},
+    Executor,
+};
+use proto;
+use proxy::node::NodeContext;
+use semver::Version;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::fmt;
+use std::str::FromStr;
+use storage::View;
+use to_handle;
+use utils::{jni_cache::runtime_adapter, panic_on_exception, unwrap_jni};
+use Handle;
+use JniErrorKind;
+use JniResult;
+
+/// A proxy for `ServiceRuntimeAdapter`s.
+#[derive(Clone)]
+pub struct JavaRuntimeProxy {
+    exec: Executor,
+    runtime_adapter: GlobalRef,
+    deployed_artifacts: HashSet<JavaArtifactId>,
+    started_services: HashMap<InstanceId, Instance>,
+    started_services_by_name: HashMap<String, InstanceId>,
+}
+
+/// Service identification properties within `JavaRuntimeProxy`
+#[derive(Debug, Clone)]
+struct Instance {
+    id: InstanceId,
+    name: String,
+}
+
+/// Artifact identification properties within `JavaRuntimeProxy`
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct JavaArtifactId {
```

Comment:
I don't see any code using these separate components. Please make it a string/a string typedef/a wrapper around it.